build_modules/lib/src/kernel_builder.dart

```diff
@@ -148,28 +143,68 @@ Future<void> _createKernel(
   }
 }
 
-/// Adds the source or kernel dependencies for [dependency] to
+/// Adds the source or kernel dependencies for [root] to
 /// [transitiveKernelDeps] or [transitiveSourceDeps].
 Future<void> _addModuleDeps(
-    Module dependency,
     Module root,
     List<AssetId> transitiveKernelDeps,
     List<AssetId> transitiveSourceDeps,
     BuildStep buildStep,
     String outputExtension) async {
-  var kernelId = dependency.primarySource.changeExtension(outputExtension);
-  if (await buildStep.canRead(kernelId)) {
-    // If we can read the kernel file, but it depends on any module in this
-    // package, then we need to only provide sources for that file since its
-    // dependencies in this package will only be providing sources as well.
-    if ((await dependency.computeTransitiveDependencies(buildStep))
-        .any((m) => m.primarySource.package == root.primarySource.package)) {
-      transitiveSourceDeps.addAll(dependency.sources);
+  final toCrawl = Queue.of([root.primarySource]);
+  final nodes = {root.primarySource: _ModuleNode(root, true)};
+
+  // Helper for any time we encounter a module which is provided via source and
+  // not kernel.
+  //
+  // For each of those we have to ensure that all parents of that module are
+  // provided via source as well.
+  void Function(AssetId) updateParents;
+  updateParents = (AssetId nodeId) {
+    var node = nodes[nodeId];
+    for (var parent in node.parents) {
+      var parentNode = nodes[parent];
+      if (!parentNode.sourceOnly) {
+        parentNode.sourceOnly = true;
+        parentNode.parents.forEach(updateParents);
+      }
+    }
+  };
+
+  // Builds the module graph, the `sourceOnly` property of nodes gets
+  // incrementally updated so we need to wait until we are done to actually
+  // provide the deps.
+  while (toCrawl.isNotEmpty) {
```

Comment:
[optional] It's plausible we could accomplish this with https://pub.dartlang.org/documentation/graphs/latest/graphs/crawlAsync.html and still crawl up the `parents` to mark them each time we find a parent relationship.

Reply:
I think it would end up feeling a bit weird - feels better to be more explicit given how specialized this is.